src/persist/src/operators/stream.rs

```diff
@@ -301,17 +329,183 @@ where
 
         (data_output_stream, error_output_stream)
     }
+
+    fn conditional_seal<K2, V2>(
+        &self,
+        name: &str,
+        condition_input: &Stream<G, ((K2, V2), u64, isize)>,
+        primary_write: StreamWriteHandle<K, V>,
+        condition_write: StreamWriteHandle<K2, V2>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    )
+    where
+        K2: TimelyData + Codec,
+        V2: TimelyData + Codec,
+    {
+        let operator_name = format!("conditional_seal({})", name);
+        let mut seal_op = OperatorBuilder::new(operator_name.clone(), self.scope());
+
+        let mut primary_data_input = seal_op.new_input(&self, Pipeline);
+        let mut condition_data_input = seal_op.new_input(condition_input, Pipeline);
+
+        let (mut data_output, data_output_stream) = seal_op.new_output();
+
+        let mut primary_data_buffer = Vec::new();
+        let mut condition_data_buffer = Vec::new();
+
+        // We only seal from one worker because sealing from multiple workers could lead to a race
+        // conditions where one worker seals up to time `t` while another worker is still trying to
+        // write data with timestamps that are not beyond `t`.
+        //
+        // Upstream persist() operators will only advance their frontier when writes are succesful.
+        // With timely progress tracking we are therefore sure that when the frontier advances for
+        // worker 0, it has advanced to at least that point for all upstream operators.
+        //
+        // Alternative solutions would be to "teach" persistence to work with seals from multiple
+        // workers, or to use a non-timely solution for keeping track of outstanding write
+        // capabilities.
+        let active_seal_operator = self.scope().index() == 0;
+
+        seal_op.build(move |mut capabilities| {
+            let mut primary_notificator = FrontierNotificator::new();
+            let mut condition_notificator = FrontierNotificator::new();
+
+            if active_seal_operator {
+                let initial_primary_cap = capabilities.pop().expect("missing capability");
+                let initial_condition_cap = initial_primary_cap.clone();
```

Comment:
I have no idea how this works, but I'd expect that `capabilities` would have one per input, is that not the case?